File: packages/sardis-sdk-python/src/sardis_sdk/resources/funding.py
"""
Funding resource for Sardis SDK.

Sardis Protocol v1.0 -- Funding commitments and cells provide a structured
way to pre-fund agent wallets. A commitment locks a total value from a vault
reference, and cells are the subdivisions that agents can draw against.

This module provides both async and sync interfaces.
"""
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .base import AsyncBaseResource, SyncBaseResource

if TYPE_CHECKING:
    import builtins
    from decimal import Decimal

    from ..client import TimeoutConfig


class AsyncFundingResource(AsyncBaseResource):
    """Async resource for funding commitments and cells.

    Funding commitments represent pre-allocated capital from a vault.
    Cells are subdivisions of a commitment that agents can draw against.

    Example:
        ```python
        async with AsyncSardisClient(api_key="...") as client:
            # Create a funding commitment
            commitment = await client.funding.create_commitment(
                vault_ref="vault_abc",
                total_value=Decimal("10000.00"),
                currency="USDC",
                cell_strategy="equal",
            )

            # List cells for a commitment
            cells = await client.funding.list_cells(
                commitment_id=commitment["id"],
            )

            # Split a cell into smaller denominations
            result = await client.funding.split_cell(
                cell_id="cell_xyz",
                amounts=[Decimal("100.00"), Decimal("200.00")],
            )
        ```
    """

    async def create_commitment(
        self,
        vault_ref: str,
        total_value: Decimal,
        currency: str = "USDC",
        cell_strategy: str = "equal",
        cell_count: int | None = None,
        cell_amounts: builtins.list[Decimal] | None = None,
        chain: str | None = None,
        metadata: dict[str, Any] | None = None,
        timeout: float | TimeoutConfig | None = None,
    ) -> dict[str, Any]:
        """Create a new funding commitment from a vault.

        Args:
            vault_ref: Reference to the source vault
            total_value: Total value to commit
            currency: Currency code (default: USDC)
            cell_strategy: How to subdivide ("equal", "custom", "single")
            cell_count: Number of cells for "equal" strategy
            cell_amounts: Explicit cell amounts for "custom" strategy
            chain: Optional chain identifier
            metadata: Optional metadata dictionary
            timeout: Optional request timeout

        Returns:
            The created funding commitment
        """
        payload: dict[str, Any] = {
            "vault_ref": vault_ref,
            "total_value": str(total_value),
            "currency": currency,
            "cell_strategy": cell_strategy,
        }

        if cell_count is not None:
            payload["cell_count"] = cell_count
        if cell_amounts is not None:
            payload["cell_amounts"] = [str(a) for a in cell_amounts]
        if chain is not None:
            payload["chain"] = chain
        if metadata is not None:
            payload["metadata"] = metadata

        return await self._post("funding/commitments", payload, timeout=timeout)

    async def list_commitments(
        self,
        status: str | None = None,
        limit: int = 100,
        timeout: float | TimeoutConfig | None = None,
    ) -> builtins.list[dict[str, Any]]:
        """List funding commitments.

        Args:
            status: Filter by status (e.g., "active", "depleted", "cancelled")
            limit: Maximum number of commitments to return
            timeout: Optional request timeout

        Returns:
            List of funding commitments
        """
        params: dict[str, Any] = {"limit": limit}
        if status is not None:
            params["status"] = status

        data = await self._get("funding/commitments", params=params, timeout=timeout)

        if isinstance(data, list):
            return data
        return data.get("commitments", data.get("items", []))

    async def list_cells(
        self,
        commitment_id: str | None = None,
        status: str | None = None,
        currency: str | None = None,
        limit: int = 100,
        timeout: float | TimeoutConfig | None = None,
    ) -> builtins.list[dict[str, Any]]:
        """List funding cells.

        Args:
            commitment_id: Filter by parent commitment ID
            status: Filter by status (e.g., "available", "locked", "spent")
            currency: Filter by currency code
            limit: Maximum number of cells to return
            timeout: Optional request timeout

        Returns:
            List of funding cells
        """
        params: dict[str, Any] = {"limit": limit}
        if commitment_id is not None:
            params["commitment_id"] = commitment_id
        if status is not None:
            params["status"] = status
        if currency is not None:
            params["currency"] = currency

        data = await self._get("funding/cells", params=params, timeout=timeout)

        if isinstance(data, list):
            return data
        return data.get("cells", data.get("items", []))

    async def split_cell(
        self,
        cell_id: str,
        amounts: builtins.list[Decimal],
        timeout: float | TimeoutConfig | None = None,
    ) -> dict[str, Any]:
        """Split a funding cell into smaller cells.

        Args:
            cell_id: The cell to split
            amounts: List of amounts for the new cells (must sum to original)
            timeout: Optional request timeout

        Returns:
            Result containing the new cells
        """
        payload: dict[str, Any] = {
            "amounts": [str(a) for a in amounts],
        }

        return await self._post(
            f"funding/cells/{cell_id}/split", payload, timeout=timeout
        )

    async def merge_cells(
        self,
        cell_ids: builtins.list[str],
        timeout: float | TimeoutConfig | None = None,
    ) -> dict[str, Any]:
        """Merge multiple funding cells into one.

        Args:
            cell_ids: List of cell IDs to merge (must share same commitment)
            timeout: Optional request timeout

        Returns:
            The merged cell
        """
        payload: dict[str, Any] = {
            "cell_ids": cell_ids,
        }

        return await self._post("funding/cells/merge", payload, timeout=timeout)


class FundingResource(SyncBaseResource):
    """Sync resource for funding commitments and cells.

    Funding commitments represent pre-allocated capital from a vault.
    Cells are subdivisions of a commitment that agents can draw against.

    Example:
        ```python
        with SardisClient(api_key="...") as client:
            # Create a funding commitment
            commitment = client.funding.create_commitment(
                vault_ref="vault_abc",
                total_value=Decimal("10000.00"),
                currency="USDC",
                cell_strategy="equal",
            )

            # List cells for a commitment
            cells = client.funding.list_cells(
                commitment_id=commitment["id"],
            )

            # Split a cell into smaller denominations
            result = client.funding.split_cell(
                cell_id="cell_xyz",
                amounts=[Decimal("100.00"), Decimal("200.00")],
            )
        ```
    """

    def create_commitment(
        self,
        vault_ref: str,
        total_value: Decimal,
        currency: str = "USDC",
        cell_strategy: str = "equal",
        cell_count: int | None = None,
        cell_amounts: builtins.list[Decimal] | None = None,
        chain: str | None = None,
        metadata: dict[str, Any] | None = None,
        timeout: float | TimeoutConfig | None = None,
    ) -> dict[str, Any]:
        """Create a new funding commitment from a vault.

        Args:
            vault_ref: Reference to the source vault
            total_value: Total value to commit
            currency: Currency code (default: USDC)
            cell_strategy: How to subdivide ("equal", "custom", "single")
            cell_count: Number of cells for "equal" strategy
            cell_amounts: Explicit cell amounts for "custom" strategy
            chain: Optional chain identifier
            metadata: Optional metadata dictionary
            timeout: Optional request timeout

        Returns:
            The created funding commitment
        """
        payload: dict[str, Any] = {
            "vault_ref": vault_ref,
            "total_value": str(total_value),
            "currency": currency,
            "cell_strategy": cell_strategy,
        }

        if cell_count is not None:
            payload["cell_count"] = cell_count
        if cell_amounts is not None:
            payload["cell_amounts"] = [str(a) for a in cell_amounts]
        if chain is not None:
            payload["chain"] = chain
        if metadata is not None:
            payload["metadata"] = metadata

        return self._post("funding/commitments", payload, timeout=timeout)

    def list_commitments(
        self,
        status: str | None = None,
        limit: int = 100,
        timeout: float | TimeoutConfig | None = None,
    ) -> builtins.list[dict[str, Any]]:
        """List funding commitments.

        Args:
            status: Filter by status (e.g., "active", "depleted", "cancelled")
            limit: Maximum number of commitments to return
            timeout: Optional request timeout

        Returns:
            List of funding commitments
        """
        params: dict[str, Any] = {"limit": limit}
        if status is not None:
            params["status"] = status

        data = self._get("funding/commitments", params=params, timeout=timeout)

        if isinstance(data, list):
            return data
        return data.get("commitments", data.get("items", []))

    def list_cells(
        self,
        commitment_id: str | None = None,
        status: str | None = None,
        currency: str | None = None,
        limit: int = 100,
        timeout: float | TimeoutConfig | None = None,
    ) -> builtins.list[dict[str, Any]]:
        """List funding cells.

        Args:
            commitment_id: Filter by parent commitment ID
            status: Filter by status (e.g., "available", "locked", "spent")
            currency: Filter by currency code
            limit: Maximum number of cells to return
            timeout: Optional request timeout

        Returns:
            List of funding cells
        """
        params: dict[str, Any] = {"limit": limit}
        if commitment_id is not None:
            params["commitment_id"] = commitment_id
        if status is not None:
            params["status"] = status
        if currency is not None:
            params["currency"] = currency

        data = self._get("funding/cells", params=params, timeout=timeout)

        if isinstance(data, list):
            return data
        return data.get("cells", data.get("items", []))

    def split_cell(
        self,
        cell_id: str,
        amounts: builtins.list[Decimal],
        timeout: float | TimeoutConfig | None = None,
    ) -> dict[str, Any]:
        """Split a funding cell into smaller cells.

        Args:
            cell_id: The cell to split
            amounts: List of amounts for the new cells (must sum to original)
            timeout: Optional request timeout

        Returns:
            Result containing the new cells
        """
        payload: dict[str, Any] = {
            "amounts": [str(a) for a in amounts],
        }

        return self._post(
            f"funding/cells/{cell_id}/split", payload, timeout=timeout
        )

    def merge_cells(
        self,
        cell_ids: builtins.list[str],
        timeout: float | TimeoutConfig | None = None,
    ) -> dict[str, Any]:
        """Merge multiple funding cells into one.

        Args:
            cell_ids: List of cell IDs to merge (must share same commitment)
            timeout: Optional request timeout

        Returns:
            The merged cell
        """
        payload: dict[str, Any] = {
            "cell_ids": cell_ids,
        }

        return self._post("funding/cells/merge", payload, timeout=timeout)


__all__ = [
    "AsyncFundingResource",
    "FundingResource",
]

File: packages/sardis-mcp-server/src/tools/funding.ts
/**
 * Sardis MCP Server - Funding Tools (Sardis Protocol v1.0)
 *
 * Funding Commitments and Cells provide granular control over wallet liquidity.
 * A Funding Commitment reserves a specific amount for a purpose, creating one or
 * more Funding Cells. Cells can be split (subdivide a budget) or merged (consolidate
 * fragmented liquidity) without moving funds on-chain.
 *
 * Tools:
 * - sardis_create_funding_commitment: Reserve funds for a specific purpose
 * - sardis_list_funding_cells: List funding cells with optional filters
 * - sardis_split_cell: Split a funding cell into smaller cells
 * - sardis_merge_cells: Merge multiple funding cells into one
 */

import { z } from 'zod';
import type { ToolDefinition, ToolHandler, ToolResult } from './types.js';
import { apiRequest } from '../api.js';
import { getConfig } from '../config.js';

// Zod schemas
const CreateFundingCommitmentSchema = z.object({
  amount: z.string().describe('Total amount to commit as a decimal string (e.g., "500.00")'),
  token: z.string().optional().default('USDC').describe('Token for the commitment (default: USDC)'),
  chain: z.string().optional().default('base').describe('Blockchain network (default: base)'),
  purpose: z.string().describe('Purpose of the funding commitment'),
  mandate_id: z.string().optional().describe('Optional spending mandate to bind this commitment to'),
  cell_count: z.number().optional().default(1).describe('Number of equal cells to create (default: 1)'),
  expires_in_hours: z.number().optional().default(720).describe('Commitment expiration in hours (default: 720 / 30 days)'),
});

const ListFundingCellsSchema = z.object({
  commitment_id: z.string().optional().describe('Filter by funding commitment ID'),
  status: z.string().optional().describe('Filter by status: available, reserved, spent, expired'),
  limit: z.number().optional().default(20).describe('Maximum results to return'),
  offset: z.number().optional().default(0).describe('Pagination offset'),
});

const SplitCellSchema = z.object({
  cell_id: z.string().describe('ID of the funding cell to split'),
  amounts: z.array(z.string()).describe('Array of amounts for the new cells (must sum to original cell amount)'),
  labels: z.array(z.string()).optional().describe('Optional labels for each new cell'),
});

const MergeCellsSchema = z.object({
  cell_ids: z.array(z.string()).min(2).describe('Array of funding cell IDs to merge (minimum 2)'),
  label: z.string().optional().describe('Optional label for the merged cell'),
});

// Tool definitions
export const fundingToolDefinitions: ToolDefinition[] = [
  {
    name: 'sardis_create_funding_commitment',
    description:
      'Create a funding commitment that reserves wallet funds for a specific purpose. ' +
      'The commitment creates one or more funding cells that represent subdivided, ' +
      'trackable portions of the reserved amount. Useful for budgeting across multiple ' +
      'agents, projects, or time periods.',
    inputSchema: {
      type: 'object',
      properties: {
        amount: {
          type: 'string',
          description: 'Total amount to commit as a decimal string (e.g., "500.00")',
        },
        token: {
          type: 'string',
          description: 'Token for the commitment (default: USDC)',
          enum: ['USDC', 'USDT', 'PYUSD', 'EURC'],
        },
        chain: {
          type: 'string',
          description: 'Blockchain network (default: base)',
        },
        purpose: {
          type: 'string',
          description: 'Purpose of the funding commitment (e.g., "Q1 marketing budget")',
        },
        mandate_id: {
          type: 'string',
          description: 'Optional spending mandate to bind this commitment to',
        },
        cell_count: {
          type: 'number',
          description: 'Number of equal cells to create (default: 1)',
        },
        expires_in_hours: {
          type: 'number',
          description: 'Commitment expiration in hours (default: 720 / 30 days)',
        },
      },
      required: ['amount', 'purpose'],
    },
  },
  {
    name: 'sardis_list_funding_cells',
    description:
      'List funding cells with optional filtering by commitment, status, or pagination. ' +
      'Funding cells are the atomic unit of reserved liquidity within a commitment.',
    inputSchema: {
      type: 'object',
      properties: {
        commitment_id: {
          type: 'string',
          description: 'Filter by funding commitment ID',
        },
        status: {
          type: 'string',
          enum: ['available', 'reserved', 'spent', 'expired'],
          description: 'Filter by cell status',
        },
        limit: {
          type: 'number',
          description: 'Maximum results to return (default: 20)',
        },
        offset: {
          type: 'number',
          description: 'Pagination offset (default: 0)',
        },
      },
      required: [],
    },
  },
  {
    name: 'sardis_split_cell',
    description:
      'Split a funding cell into multiple smaller cells. The amounts must sum exactly to ' +
      'the original cell amount. Useful for subdividing a budget across multiple agents ' +
      'or purposes without moving funds on-chain.',
    inputSchema: {
      type: 'object',
      properties: {
        cell_id: {
          type: 'string',
          description: 'ID of the funding cell to split',
        },
        amounts: {
          type: 'array',
          items: { type: 'string' },
          description: 'Array of amounts for the new cells (must sum to original cell amount)',
        },
        labels: {
          type: 'array',
          items: { type: 'string' },
          description: 'Optional labels for each new cell',
        },
      },
      required: ['cell_id', 'amounts'],
    },
  },
  {
    name: 'sardis_merge_cells',
    description:
      'Merge multiple funding cells into a single cell. All cells must belong to the same ' +
      'commitment and be in "available" status. Useful for consolidating fragmented ' +
      'liquidity back into a single budget line.',
    inputSchema: {
      type: 'object',
      properties: {
        cell_ids: {
          type: 'array',
          items: { type: 'string' },
          description: 'Array of funding cell IDs to merge (minimum 2)',
        },
        label: {
          type: 'string',
          description: 'Optional label for the merged cell',
        },
      },
      required: ['cell_ids'],
    },
  },
];

// Tool handlers
export const fundingToolHandlers: Record<string, ToolHandler> = {
  sardis_create_funding_commitment: async (args: unknown): Promise<ToolResult> => {
    const parsed = CreateFundingCommitmentSchema.safeParse(args);
    if (!parsed.success) {
      return {
        content: [{ type: 'text', text: `Invalid request: ${parsed.error.message}` }],
        isError: true,
      };
    }

    const { amount, token, chain, purpose, mandate_id, cell_count, expires_in_hours } = parsed.data;
    const config = getConfig();

    if (!config.apiKey || config.mode === 'simulated') {
      const simId = `fc_sim_${Date.now().toString(36)}`;
      const cellAmount = (parseFloat(amount) / cell_count).toFixed(2);
      const cells = Array.from({ length: cell_count }, (_, i) => ({
        cell_id: `cell_sim_${Date.now().toString(36)}_${i}`,
        amount: cellAmount,
        status: 'available',
      }));
      const expiresAt = new Date(Date.now() + expires_in_hours * 3600_000).toISOString();

      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(
              {
                success: true,
                commitment_id: simId,
                amount,
                token,
                chain,
                purpose,
                mandate_id: mandate_id || null,
                cells,
                expires_at: expiresAt,
                created_at: new Date().toISOString(),
                message: `Funding commitment created: ${amount} ${token} split into ${cell_count} cell(s)`,
              },
              null,
              2,
            ),
          },
        ],
      };
    }

    try {
      const result = await apiRequest<Record<string, unknown>>(
        'POST',
        '/api/v2/funding/commitments',
        { amount, token, chain, purpose, mandate_id, cell_count, expires_in_hours },
      );

      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(
              {
                success: true,
                ...result,
                message: `Funding commitment ${result.commitment_id} created (${amount} ${token})`,
              },
              null,
              2,
            ),
          },
        ],
      };
    } catch (error) {
      return {
        content: [
          { type: 'text', text: JSON.stringify({ success: false, error: error instanceof Error ? error.message : 'Failed to create funding commitment' }, null, 2) },
        ],
        isError: true,
      };
    }
  },

  sardis_list_funding_cells: async (args: unknown): Promise<ToolResult> => {
    const parsed = ListFundingCellsSchema.safeParse(args);
    const { commitment_id, status, limit, offset } = parsed.success
      ? parsed.data
      : { commitment_id: undefined, status: undefined, limit: 20, offset: 0 };

    const config = getConfig();

    if (!config.apiKey || config.mode === 'simulated') {
      const simTime = new Date().toISOString();
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(
              {
                cells: [
                  {
                    id: `cell_sim_${Date.now().toString(36)}`,
                    commitment_id: commitment_id || `fc_sim_example`,
                    amount: '100.00',
                    token: 'USDC',
                    status: status || 'available',
                    label: null,
                    created_at: simTime,
                  },
                ],
                total: 1,
              },
              null,
              2,
            ),
          },
        ],
      };
    }

    try {
      const params = new URLSearchParams({
        limit: limit.toString(),
        offset: offset.toString(),
      });
      if (commitment_id) params.append('commitment_id', commitment_id);
      if (status) params.append('status', status);

      const result = await apiRequest<{ cells: Record<string, unknown>[]; total: number }>(
        'GET',
        `/api/v2/funding/cells?${params}`,
      );

      return {
        content: [{ type: 'text', text: JSON.stringify(result, null, 2) }],
      };
    } catch (error) {
      return {
        content: [
          { type: 'text', text: `Failed to list funding cells: ${error instanceof Error ? error.message : 'Unknown error'}` },
        ],
        isError: true,
      };
    }
  },

  sardis_split_cell: async (args: unknown): Promise<ToolResult> => {
    const parsed = SplitCellSchema.safeParse(args);
    if (!parsed.success) {
      return {
        content: [{ type: 'text', text: `Invalid request: ${parsed.error.message}` }],
        isError: true,
      };
    }

    const { cell_id, amounts, labels } = parsed.data;
    const config = getConfig();

    if (!config.apiKey || config.mode === 'simulated') {
      const newCells = amounts.map((amt, i) => ({
        cell_id: `cell_sim_${Date.now().toString(36)}_${i}`,
        amount: amt,
        label: labels?.[i] || null,
        status: 'available',
      }));

      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(
              {
                success: true,
                original_cell_id: cell_id,
                new_cells: newCells,
                message: `Cell ${cell_id} split into ${amounts.length} cells`,
              },
              null,
              2,
            ),
          },
        ],
      };
    }

    try {
      const result = await apiRequest<Record<string, unknown>>(
        'POST',
        `/api/v2/funding/cells/${cell_id}/split`,
        { amounts, labels },
      );

      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(
              { success: true, ...result, message: `Cell ${cell_id} split into ${amounts.length} cells` },
              null,
              2,
            ),
          },
        ],
      };
    } catch (error) {
      return {
        content: [
          { type: 'text', text: JSON.stringify({ success: false, error: error instanceof Error ? error.message : 'Failed to split cell' }, null, 2) },
        ],
        isError: true,
      };
    }
  },

  sardis_merge_cells: async (args: unknown): Promise<ToolResult> => {
    const parsed = MergeCellsSchema.safeParse(args);
    if (!parsed.success) {
      return {
        content: [{ type: 'text', text: `Invalid request: ${parsed.error.message}` }],
        isError: true,
      };
    }

    const { cell_ids, label } = parsed.data;
    const config = getConfig();

    if (!config.apiKey || config.mode === 'simulated') {
      const mergedId = `cell_sim_merged_${Date.now().toString(36)}`;
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(
              {
                success: true,
                merged_cell_id: mergedId,
                source_cell_ids: cell_ids,
                amount: '200.00',
                label: label || null,
                status: 'available',
                message: `${cell_ids.length} cells merged into ${mergedId}`,
              },
              null,
              2,
            ),
          },
        ],
      };
    }

    try {
      const result = await apiRequest<Record<string, unknown>>(
        'POST',
        '/api/v2/funding/cells/merge',
        { cell_ids, label },
      );

      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(
              { success: true, ...result, message: `${cell_ids.length} cells merged successfully` },
              null,
              2,
            ),
          },
        ],
      };
    } catch (error) {
      return {
        content: [
          { type: 'text', text: JSON.stringify({ success: false, error: error instanceof Error ? error.message : 'Failed to merge cells' }, null, 2) },
        ],
        isError: true,
      };
    }
  },
};
